moped-etl/arcgis/components_to_agol.py

```diff
@@ -100,28 +110,112 @@ def main():
-        delete_features(feature_type)
 
-        logger.info(
-            f"Uploading {len(features)} features in chunks of {UPLOAD_CHUNK_SIZE}..."
-        )
-        for feature_chunk in chunks(features, UPLOAD_CHUNK_SIZE):
-            logger.info("Uploading chunk....")
-            add_features(feature_type, feature_chunk)
+def main(args):
+    logger.info("Getting token...")
+    get_token()
+
+    if args.full:
+        logger.info("Getting component features from all projects...")
+        data = make_hasura_request(
+            query=COMPONENTS_QUERY_BY_LAST_UPDATE_DATE,
+            variables={"where": {}},
+        )["component_arcgis_online_view"]
+
+        all_features = make_all_features(data)
+
+        for feature_type in ["points", "lines", "combined"]:
+            logger.info(f"Processing {feature_type} features...")
+            features = all_features[feature_type]
+
+            logger.info("Deleting all existing features...")
+            delete_all_features(feature_type)
+
+            logger.info(
+                f"Uploading {len(features)} features in chunks of {UPLOAD_CHUNK_SIZE}..."
+            )
+            for feature_chunk in chunks(features, UPLOAD_CHUNK_SIZE):
+                logger.info("Uploading chunk....")
+                add_features(feature_type, feature_chunk)
+    else:
+        logger.info("Getting component features from recently updated projects...")
+        data = make_hasura_request(
+            query=COMPONENTS_QUERY_BY_LAST_UPDATE_DATE,
+            variables={"where": {"project_updated_at": {"_gt": args.date}}},
+        )["component_arcgis_online_view"]
+
+        all_features = make_all_features(data)
+
+        # Get unique project IDs that need to have features deleted & replaced
+        project_ids = []
+
+        for component in data:
+            project_ids.append(component["project_id"])
+
+        project_ids_for_feature_delete = list(set(project_ids))
+
+        # Delete outdated feature from AGOL and add updated features
+        for feature_type in ["points", "lines", "combined"]:
+            logger.info(f"Processing {feature_type} features...")
+            logger.info(
+                f"Deleting all existing features in {feature_type} layer for updated projects..."
+            )
+            for delete_chunk in chunks(
+                project_ids_for_feature_delete, UPLOAD_CHUNK_SIZE
+            ):
+                joined_project_ids = ", ".join(str(x) for x in delete_chunk)
+                logger.info(
+                    f"Deleting {len(project_ids_for_feature_delete)} features with project ids {joined_project_ids} in chunks of {UPLOAD_CHUNK_SIZE}..."
+                )
+                delete_features_by_project_ids(feature_type, delete_chunk)
+
+            features = all_features[feature_type]
+
+            logger.info(
+                f"Uploading {len(features)} features in chunks of {UPLOAD_CHUNK_SIZE}..."
```

Comment:
the console logging was throwing me off a bit here, because i encountered this:

```shell
 components-to-agol.INFO: Deleting 3 features with project ids 3392, 786, 3523 in chunks of 100...
 components-to-agol.INFO: Uploading 2 features in chunks of 100...
```

and i see that in certain cases there is a difference between the length of `project_ids_for_feature_delete` vs `features`—because `project_ids_for_feature_delete` includes IDs from all feature types (point/line/combined) while the `features` array is specific to the current `feature_type`. 

it's definitely not a bug, but i think you could make the AGOL deletes slightly more optimal by parsing the project IDs to delete out of the `features` list. i'm not sure though—i've confused myself by writing this comment 😛

Reply:
it is confusing! but this makes sense to me and a result of my rushed effort to improve logging. 😅 I'll iron this out.

Reply:
I found it confusing too, trying to figure out what was happening, but then just trusted the script and moved on 😅